smithy-model/src/main/java/software/amazon/smithy/model/knowledge/TextIndex.java

```diff
@@ -0,0 +1,133 @@
+/*
+ * Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.smithy.model.knowledge;
+
+import java.util.ArrayDeque;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Deque;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import software.amazon.smithy.model.Model;
+import software.amazon.smithy.model.loader.Prelude;
+import software.amazon.smithy.model.node.Node;
+import software.amazon.smithy.model.node.ObjectNode;
+import software.amazon.smithy.model.shapes.MemberShape;
+import software.amazon.smithy.model.shapes.Shape;
+import software.amazon.smithy.model.traits.ReferencesTrait;
+import software.amazon.smithy.model.traits.Trait;
+import software.amazon.smithy.model.validation.validators.TraitValueValidator;
+import software.amazon.smithy.utils.SmithyUnstableApi;
+
+@SmithyUnstableApi
+public final class TextIndex implements KnowledgeIndex {
+    private final List<TextInstance> textInstanceList = new ArrayList<>();
+
+    public TextIndex(Model model) {
+        Set<String> visitedNamespaces = new HashSet<>();
+        // Validating the prelude is a feature for internal-only Smithy development
+        Node validatePreludeNode = model.getMetadata().get(TraitValueValidator.VALIDATE_PRELUDE);
+        boolean validatePrelude = validatePreludeNode != null
+                ? validatePreludeNode.expectBooleanNode().getValue()
+                : false;
+
+        for (final Shape shape : model.toSet()) {
+            if (validatePrelude || !Prelude.isPreludeShape(shape)) {
+                if (visitedNamespaces.add(shape.getId().getNamespace())) {
+                    textInstanceList.add(TextInstance.createNamespaceText(shape.getId().getNamespace()));
+                }
+                computeShapeTextInstances(shape, textInstanceList, model);
+            }
+        }
+    }
+
+    public static TextIndex of(Model model) {
+        return model.getKnowledge(TextIndex.class, TextIndex::new);
+    }
+
+    public Collection<TextInstance> getTextInstances() {
+        return Collections.unmodifiableList(textInstanceList);
+    }
+
+    private static void computeShapeTextInstances(
+            Shape shape,
+            Collection<TextInstance> textInstances,
+            Model model) {
```

Comment:
nit: We usually put the `) {` for multi-line declarations on the next line, inline with the start of the declaration. (Others in the file as well.)